libraries/Filter/HarmonicNotchFilter.cpp

```diff
@@ -0,0 +1,193 @@
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "HarmonicNotchFilter.h"
+
+#define HNF_MAX_FILTERS 3
+#define HNF_MAX_HARMONICS 8
+
+// table of user settable parameters
+const AP_Param::GroupInfo HarmonicNotchFilterParams::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Harmonic Notch Filter enable
+    // @Description: Harmonic Notch Filter enable
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLE", 1, HarmonicNotchFilterParams, _enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: FREQ
+    // @DisplayName: Harmonic Notch Filter base frequency
+    // @Description: Harmonic Notch Filter base center frequency in Hz
+    // @Range: 10 400
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("FREQ", 2, HarmonicNotchFilterParams, _center_freq_hz, 80),
+
+    // @Param: BW
+    // @DisplayName: Harmonic Notch Filter bandwidth
+    // @Description: Harmonic Notch Filter bandwidth in Hz
+    // @Range: 5 100
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("BW", 3, HarmonicNotchFilterParams, _bandwidth_hz, 20),
+
+    // @Param: ATT
+    // @DisplayName: Harmonic Notch Filter attenuation
+    // @Description: Harmonic Notch Filter attenuation in dB
+    // @Range: 5 30
+    // @Units: dB
+    // @User: Advanced
+    AP_GROUPINFO("ATT", 4, HarmonicNotchFilterParams, _attenuation_dB, 15),
+
+    // @Param: HMNCS
+    // @DisplayName: Harmonic Notch Filter harmonics
+    // @Description: Bitmask of harmonic frequencies to apply Harmonic Notch Filter to. This option takes effect on the next reboot. A maximum of 3 harmonics can be used at any one time
+    // @Bitmask: 0:1st harmonic,1:2nd harmonic,2:3rd harmonic,3:4th hamronic,4:5th harmonic,5:6th harmonic,6:7th harmonic,7:8th harmonic
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("HMNCS", 5, HarmonicNotchFilterParams, _harmonics, 3),
+
+    // @Param: REF
+    // @DisplayName: Harmonic Notch Filter reference value
+    // @Description: Reference value associated with the specified frequency to facilitate frequency scaling of the Harmonic Notch Filter
+    // @User: Advanced
+    // @Range: 0.1 0.9
+    // @RebootRequired: True
+    AP_GROUPINFO("REF", 6, HarmonicNotchFilterParams, _reference, 0.1f),
+
+    AP_GROUPEND
+};
+
+/*
+  destroy all of the associated notch filters
+ */
+template <class T>
+HarmonicNotchFilter<T>::~HarmonicNotchFilter() {
+    delete[] _filters;
+    _num_filters = 0;
+}
+
+/*
+  initialise the associated filters with the provided shaping constraints
+  the constraints are used to determine attenuation (A) and quality (Q) factors for the filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, float bandwidth_hz, float attenuation_dB)
+{
+    if (_filters == nullptr) {
```

Comment:
as far as I can tell this is only called once, and it is called before _filters is allocated, so this function never sets _initialised to true, so the filter is never activated.

Reply:
It's called repeatedly in AP_InertialSensor_Backend::update_gyro() if the settings change. allocate_filters() is called in AP_InertialSensor::init() which is called before the scheduler has started calling update (even though the backends are allocated earlier). That said in my testing I am updating the center_freq repeatedly, so if there was an issue I wouldn't see it - let me double check in SITL.

Reply:
You are completely correct - fixed now!

Reply:
thanks, I do wonder how much of the testing this invalidates though. The filter didn't do anything when I tested it

Reply:
the irony is that it was one of the many reviews that introduced the problem. So far the main source of in-flight testing has been me. So I have done some basic aggressive flying testing with the fixes below but not acro-based testing that I was doing before - I will do this again when the weather improves. For CPU testing, what is the procedure there?

Reply:
Are averaged timestamps ok? I've done that successfully for my FFT change. Do you have a ballpark latency in mind? I have F4 (omnibusf4pro, pixracer and auavx2) and F7 (kakute) boards to test on.